src/DotVVM.Framework/Diagnostics/CompilationPageApiPresenter.cs

```diff
@@ -0,0 +1,40 @@
+﻿using System.Threading.Tasks;
+using DotVVM.Framework.Compilation;
+using DotVVM.Framework.Configuration;
+using DotVVM.Framework.Hosting;
+using Newtonsoft.Json;
+
+namespace DotVVM.Framework.Diagnostics
+{
+    internal class CompilationPageApiPresenter : IDotvvmPresenter
+    {
+        private readonly IDotvvmViewCompilationService compilationService;
+
+        public CompilationPageApiPresenter(IDotvvmViewCompilationService compilationService)
+        {
+            this.compilationService = compilationService;
+        }
+
+        public async Task ProcessRequest(IDotvvmRequestContext context)
+        {
+            var response = context.HttpContext.Response;
+            var isAuthorized = await context.Configuration.Diagnostics.CompilationPage.AuthorizationPredicate(context);
+            if (!isAuthorized)
+            {
+                response.StatusCode = 403;
```

Comment:
```suggestion
                response.StatusCode = 401;
```

Not very sure, but 401 means "not logged in" while 403 is "logged in, but without rights". In this case it's most likely going to be the first case.

Reply:
Yep, but 401 requires the following:

> The server generating a 401 response MUST send a WWW-Authenticate header field1 containing at least one challenge applicable to the target resource. [1]

We cannot enforce this with the authorization predicate. A 403 should mean that the server understands the requests but refuses to authorize it, which seems applicable in this situation.

\[1]: https://httpstatuses.com/401

Reply:
I think 403 is correct in this case. If the compilation page is configured to be available only on localhost without any authentication, then I believe the correct response is 403 because the server just refuses to authorize the request, not authenticate it.